clients/src/main/java/org/apache/kafka/clients/consumer/internals/MemberAssignmentReconciler.java

```diff
@@ -0,0 +1,291 @@
+        log.debug("Preparing to invoke {} with the following partitions: {}", operation.methodName, diffPartitions);
+        RebalanceCallbackEvent event = eventGenerator.generate(Collections.unmodifiableList(new ArrayList<>(diffPartitions)));
+        inflightCallback = Optional.of(event);
+
+        // Enqueue it in our background->application shared queue. This should be invoked in the Consumer.poll() method.
+        backgroundEventQueue.add(event);
+
+        // When the core event is completed in the application via Consumer.poll(), we want to invoke our code to
+        // "commit" the change.
+        event.future().whenComplete((result, error) -> {
+            if (error != null) {
+                log.warn("An error occurred when invoking {} with the following partitions: {}", operation.methodName, diffPartitions, error);
+                // TODO: should we proceed or abort?
+            }
+
+            log.debug("{} was successfully executed with the following partitions: {}", operation.methodName, diffPartitions);
+
+            // Make a copy of the current assignment set (just to be paranoid) and then modify it. The "modification"
+            // here will be different for a revocation vs. an assignment.
+            Set<TopicPartition> newAssignment = new HashSet<>(subscriptions.assignedPartitions());
+            assignmentModifier.modify(newAssignment, diffPartitions);
+
+            // And finally... assign the new set of partitions!
+            subscriptions.assignFromSubscribed(newAssignment);
+        });
+
+        return ReconciliationResult.IN_PROGRESS;
+    }
+
+    Optional<ReconciliationResult> checkInflightStatus(Timer timer) {
+        // There's no active callback we're waiting on, so proceed!
+        if (!inflightCallback.isPresent())
+            return Optional.empty();
+
+        // We don't actually need the _result_ of the event, just to know that it's complete.
+        boolean isComplete = inflightCallback.get().future().isDone();
+        boolean isExpired = timer.isExpired();
+
+        if (isComplete) {
+            // This is the happy path--we completed the callback. Clear out our inflight callback first, though.
+            inflightCallback = Optional.empty();
+            return Optional.of(ReconciliationResult.COMPLETED);
+        } else if (isExpired) {
+            // This is... not so happy. The callback did not complete in time. Still, clear out our inflight callback
+            // before returning.
+            inflightCallback = Optional.empty();
+            return Optional.of(ReconciliationResult.EXPIRED);
+        } else {
+            // So we have a callback event out there, but it's neither complete nor expired, so wait.
+            return Optional.of(ReconciliationResult.IN_PROGRESS);
+        }
+    }
+
+    static SortedSet<TopicPartition> sortPartitions(Set<TopicPartition> topicPartitions) {
```

Comment:
Do we actually need to sort the partition? I've seen that several times in the current code, but I never understand to reason for sorting.